File: src/history.py
import json
import os
from datetime import datetime
from typing import List, Dict, Optional


class HistoryManager:
    """Manages translation history storage and retrieval."""
    
    def __init__(self, history_dir: str):
        self.history_dir = history_dir
        self.history_file = os.path.join(history_dir, 'translations.json')
        os.makedirs(history_dir, exist_ok=True)
        self._ensure_history_file()
    
    def _ensure_history_file(self):
        """Create history file if it doesn't exist."""
        if not os.path.exists(self.history_file):
            with open(self.history_file, 'w', encoding='utf-8') as f:
                json.dump([], f)
    
    def add_entry(
        self,
        source_text: str,
        translated_text: str,
        source_language: str,
        target_language: str,
        mode: str = 'text',
        filename: Optional[str] = None,
        page_number: Optional[int] = None,
        summary: Optional[str] = None
    ) -> Dict:
        """Add a new translation entry to history."""
        entry = {
            'id': self._generate_id(),
            'timestamp': datetime.now().isoformat(),
            'source_text': source_text[:500],  # Store first 500 chars for preview
            'translated_text': translated_text[:500],
            'source_language': source_language,
            'target_language': target_language,
            'mode': mode,
            'filename': filename,
            'page_number': page_number,
            'summary': summary,
            'source_length': len(source_text),
            'translated_length': len(translated_text)
        }
        
        # Load existing history
        history = self._load_history()
        
        # Add new entry at the beginning
        history.insert(0, entry)
        
        # Keep only last 100 entries
        history = history[:100]
        
        # Save updated history
        self._save_history(history)
        
        return entry
    
    def get_history(self, limit: int = 50, skip: int = 0) -> List[Dict]:
        """Get history entries with pagination."""
        history = self._load_history()
        return history[skip:skip + limit]
    
    def get_entry(self, entry_id: str) -> Optional[Dict]:
        """Get a specific history entry by ID."""
        history = self._load_history()
        for entry in history:
            if entry.get('id') == entry_id:
                return entry
        return None
    
    def delete_entry(self, entry_id: str) -> bool:
        """Delete a history entry by ID."""
        history = self._load_history()
        original_length = len(history)
        history = [entry for entry in history if entry.get('id') != entry_id]
        
        if len(history) < original_length:
            self._save_history(history)
            return True
        return False
    
    def clear_history(self) -> bool:
        """Clear all history entries."""
        try:
            self._save_history([])
            return True
        except Exception:
            return False
    
    def search_history(self, query: str, limit: int = 20) -> List[Dict]:
        """Search history entries by text content."""
        history = self._load_history()
        query_lower = query.lower()
        
        results = []
        for entry in history:
            source = entry.get('source_text', '').lower()
            translated = entry.get('translated_text', '').lower()
            
            if query_lower in source or query_lower in translated:
                results.append(entry)
                
                if len(results) >= limit:
                    break
        
        return results
    
    def get_stats(self) -> Dict:
        """Get statistics about translation history."""
        history = self._load_history()
        
        if not history:
            return {
                'total_translations': 0,
                'languages': {},
                'modes': {},
                'total_characters_translated': 0
            }
        
        languages = {}
        modes = {}
        total_chars = 0
        
        for entry in history:
            # Count language pairs
            lang_pair = f"{entry.get('source_language', 'unknown')} → {entry.get('target_language', 'unknown')}"
            languages[lang_pair] = languages.get(lang_pair, 0) + 1
            
            # Count modes
            mode = entry.get('mode', 'unknown')
            modes[mode] = modes.get(mode, 0) + 1
            
            # Sum characters
            total_chars += entry.get('translated_length', 0)
        
        return {
            'total_translations': len(history),
            'languages': languages,
            'modes': modes,
            'total_characters_translated': total_chars
        }
    
    def _load_history(self) -> List[Dict]:
        """Load history from file."""
        try:
            with open(self.history_file, 'r', encoding='utf-8') as f:
                return json.load(f)
        except (json.JSONDecodeError, FileNotFoundError):
            return []
    
    def _save_history(self, history: List[Dict]):
        """Save history to file."""
        with open(self.history_file, 'w', encoding='utf-8') as f:
            json.dump(history, f, ensure_ascii=False, indent=2)
    
    def _generate_id(self) -> str:
        """Generate a unique ID for history entry."""
        import hashlib
        timestamp = datetime.now().isoformat()
        return hashlib.md5(timestamp.encode()).hexdigest()[:12]

File: static/js/history.js
// History management functionality

const History = {
    entries: [],
    currentView: 'list', // 'list' or 'stats'
    searchQuery: '',
    
    async loadHistory(limit = 50, skip = 0) {
        try {
            const url = this.searchQuery 
                ? `/api/history?limit=${limit}&skip=${skip}&q=${encodeURIComponent(this.searchQuery)}`
                : `/api/history?limit=${limit}&skip=${skip}`;
            
            const response = await fetch(url);
            const data = await response.json();
            
            if (data.success) {
                this.entries = data.entries;
                this.renderHistory();
            } else {
                showStatus('Failed to load history', 'error');
            }
        } catch (error) {
            console.error('Error loading history:', error);
            showStatus('Failed to load history', 'error');
        }
    },
    
    async loadStats() {
        try {
            const response = await fetch('/api/history/stats');
            const data = await response.json();
            
            if (data.success) {
                this.renderStats(data.stats);
            } else {
                showStatus('Failed to load statistics', 'error');
            }
        } catch (error) {
            console.error('Error loading stats:', error);
            showStatus('Failed to load statistics', 'error');
        }
    },
    
    async deleteEntry(entryId) {
        if (!confirm('Delete this translation from history?')) {
            return;
        }
        
        try {
            const response = await fetch(`/api/history/${entryId}`, {
                method: 'DELETE'
            });
            const data = await response.json();
            
            if (data.success) {
                showStatus('Entry deleted', 'success');
                this.loadHistory();
            } else {
                showStatus('Failed to delete entry', 'error');
            }
        } catch (error) {
            console.error('Error deleting entry:', error);
            showStatus('Failed to delete entry', 'error');
        }
    },
    
    async clearHistory() {
        if (!confirm('Clear all translation history? This cannot be undone.')) {
            return;
        }
        
        try {
            const response = await fetch('/api/history/clear', {
                method: 'POST'
            });
            const data = await response.json();
            
            if (data.success) {
                showStatus('History cleared', 'success');
                this.entries = [];
                this.renderHistory();
            } else {
                showStatus('Failed to clear history', 'error');
            }
        } catch (error) {
            console.error('Error clearing history:', error);
            showStatus('Failed to clear history', 'error');
        }
    },
    
    searchHistory(query) {
        this.searchQuery = query.trim();
        this.loadHistory();
    },
    
    formatTimestamp(isoString) {
        const date = new Date(isoString);
        const now = new Date();
        const diffMs = now - date;
        const diffMins = Math.floor(diffMs / 60000);
        const diffHours = Math.floor(diffMs / 3600000);
        const diffDays = Math.floor(diffMs / 86400000);
        
        if (diffMins < 1) return 'Just now';
        if (diffMins < 60) return `${diffMins} min${diffMins > 1 ? 's' : ''} ago`;
        if (diffHours < 24) return `${diffHours} hour${diffHours > 1 ? 's' : ''} ago`;
        if (diffDays < 7) return `${diffDays} day${diffDays > 1 ? 's' : ''} ago`;
        
        return date.toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' });
    },
    
    getModeIcon(mode) {
        const icons = {
            'text': '📝',
            'page': '📄',
            'document': '📚'
        };
        return icons[mode] || '📄';
    },
    
    renderHistory() {
        const container = document.getElementById('historyList');
        if (!container) return;
        
        if (this.entries.length === 0) {
            container.innerHTML = `
                <div class="empty-state centered">
                    <strong>No history yet</strong>
                    <span>Your translation history will appear here</span>
                </div>
            `;
            return;
        }
        
        container.innerHTML = this.entries.map(entry => `
            <div class="history-entry" data-entry-id="${entry.id}">
                <div class="history-entry-header">
                    <span class="history-mode">${this.getModeIcon(entry.mode)} ${entry.mode}</span>
                    <span class="history-timestamp">${this.formatTimestamp(entry.timestamp)}</span>
                    <button class="history-delete-btn icon-btn" onclick="History.deleteEntry('${entry.id}')" aria-label="Delete entry">×</button>
                </div>
                <div class="history-languages">
                    <span class="lang-badge">${entry.source_language}</span>
                    <span class="arrow">→</span>
                    <span class="lang-badge">${entry.target_language}</span>
                    ${entry.filename ? `<span class="history-filename">📎 ${escapeHtml(entry.filename)}</span>` : ''}
                    ${entry.page_number ? `<span class="history-page">Page ${entry.page_number}</span>` : ''}
                </div>
                <div class="history-content">
                    <div class="history-text">
                        <strong>Source</strong>
                        <p>${escapeHtml(entry.source_text.substring(0, 150))}${entry.source_length > 150 ? '...' : ''}</p>
                    </div>
                    <div class="history-text">
                        <strong>Translation</strong>
                        <p>${escapeHtml(entry.translated_text.substring(0, 150))}${entry.translated_length > 150 ? '...' : ''}</p>
                    </div>
                </div>
                <div class="history-actions">
                    <button class="ghost-btn small" onclick="History.copyToClipboard('${entry.id}', 'source')">Copy source</button>
                    <button class="ghost-btn small" onclick="History.copyToClipboard('${entry.id}', 'translation')">Copy translation</button>
                </div>
            </div>
        `).join('');
    },
    
    renderStats(stats) {
        const container = document.getElementById('historyStats');
        if (!container) return;
        
        const languagePairs = Object.entries(stats.languages || {})
            .sort((a, b) => b[1] - a[1])
            .slice(0, 5)
            .map(([pair, count]) => `
                <div class="stat-row">
                    <span>${escapeHtml(pair)}</span>
                    <strong>${count}</strong>
                </div>
            `).join('');
        
        const modes = Object.entries(stats.modes || {})
            .map(([mode, count]) => `
                <div class="stat-badge">
                    <span>${this.getModeIcon(mode)} ${mode}</span>
                    <strong>${count}</strong>
                </div>
            `).join('');
        
        container.innerHTML = `
            <div class="stats-grid">
                <div class="stat-card">
                    <p class="eyebrow">Total translations</p>
                    <h2>${stats.total_translations || 0}</h2>
                </div>
                <div class="stat-card">
                    <p class="eyebrow">Characters translated</p>
                    <h2>${(stats.total_characters_translated || 0).toLocaleString()}</h2>
                </div>
            </div>
            
            <div class="stat-section">
                <h3>Translation modes</h3>
                <div class="stat-badges">
                    ${modes || '<span>No data</span>'}
                </div>
            </div>
            
            <div class="stat-section">
                <h3>Top language pairs</h3>
                <div class="stat-list">
                    ${languagePairs || '<span>No data</span>'}
                </div>
            </div>
        `;
    },
    
    async copyToClipboard(entryId, type) {
        const entry = this.entries.find(e => e.id === entryId);
        if (!entry) return;
        
        const text = type === 'source' ? entry.source_text : entry.translated_text;
        
        try {
            await navigator.clipboard.writeText(text);
            showStatus(`${type === 'source' ? 'Source' : 'Translation'} copied to clipboard`, 'success');
        } catch (error) {
            console.error('Failed to copy:', error);
            showStatus('Failed to copy to clipboard', 'error');
        }
    },
    
    switchView(view) {
        this.currentView = view;
        
        // Update UI
        document.querySelectorAll('[data-history-view]').forEach(btn => {
            btn.classList.toggle('active', btn.dataset.historyView === view);
        });
        
        const listContainer = document.getElementById('historyListView');
        const statsContainer = document.getElementById('historyStatsView');
        
        if (view === 'list') {
            if (listContainer) listContainer.classList.remove('hidden');
            if (statsContainer) statsContainer.classList.add('hidden');
            this.loadHistory();
        } else {
            if (listContainer) listContainer.classList.add('hidden');
            if (statsContainer) statsContainer.classList.remove('hidden');
            this.loadStats();
        }
    },
    
    init() {
        // Set up event listeners
        const searchInput = document.getElementById('historySearchInput');
        if (searchInput) {
            let searchTimeout;
            searchInput.addEventListener('input', (e) => {
                clearTimeout(searchTimeout);
                searchTimeout = setTimeout(() => {
                    this.searchHistory(e.target.value);
                }, 300);
            });
        }
        
        const viewToggle = document.getElementById('historyViewToggle');
        if (viewToggle) {
            viewToggle.addEventListener('click', () => {
                const newView = this.currentView === 'list' ? 'stats' : 'list';
                this.switchView(newView);
                
                // Update button text
                const icon = document.getElementById('historyViewIcon');
                const label = document.getElementById('historyViewLabel');
                if (icon && label) {
                    if (newView === 'stats') {
                        icon.textContent = '📋';
                        label.textContent = 'List';
                    } else {
                        icon.textContent = '📊';
                        label.textContent = 'Stats';
                    }
                }
            });
        }
        
        const clearBtn = document.getElementById('clearHistoryBtn');
        if (clearBtn) {
            clearBtn.addEventListener('click', () => this.clearHistory());
        }
        
        // Load initial history
        this.loadHistory();
    }
};

// Initialize when history mode is activated
function initHistoryMode() {
    History.init();
}
